service/src/main/java/io/camunda/service/AdHocSubprocessActivityServices.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.service;
+
+import io.camunda.search.entities.AdHocSubprocessActivityEntity;
+import io.camunda.search.entities.AdHocSubprocessActivityEntity.ActivityType;
+import io.camunda.search.entities.ProcessDefinitionEntity;
+import io.camunda.search.exception.NotFoundException;
+import io.camunda.search.query.AdHocSubprocessActivityQuery;
+import io.camunda.search.query.SearchQueryResult;
+import io.camunda.security.auth.Authentication;
+import io.camunda.service.security.SecurityContextProvider;
+import io.camunda.zeebe.broker.client.api.BrokerClient;
+import io.camunda.zeebe.model.bpmn.Bpmn;
+import io.camunda.zeebe.model.bpmn.BpmnModelInstance;
+import io.camunda.zeebe.model.bpmn.instance.AdHocSubProcess;
+import io.camunda.zeebe.model.bpmn.instance.FlowNode;
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.stream.Collectors;
+import org.camunda.bpm.model.xml.instance.ModelElementInstance;
+
+public class AdHocSubprocessActivityServices extends ApiServices<AdHocSubprocessActivityServices> {
+
+  private final ProcessDefinitionServices processDefinitionServices;
+
+  public AdHocSubprocessActivityServices(
+      final BrokerClient brokerClient,
+      final SecurityContextProvider securityContextProvider,
+      final ProcessDefinitionServices processDefinitionServices,
+      final Authentication authentication) {
+    super(brokerClient, securityContextProvider, authentication);
+    this.processDefinitionServices = processDefinitionServices;
+  }
+
+  @Override
+  public AdHocSubprocessActivityServices withAuthentication(final Authentication authentication) {
+    return new AdHocSubprocessActivityServices(
+        brokerClient,
+        securityContextProvider,
+        processDefinitionServices.withAuthentication(authentication),
+        authentication);
+  }
+
+  public SearchQueryResult<AdHocSubprocessActivityEntity> search(
+      final AdHocSubprocessActivityQuery query) {
+    final ProcessDefinitionEntity processDefinitionEntity =
+        processDefinitionServices.getByKey(query.filter().processDefinitionKey());
+
+    final BpmnModelInstance modelInstance =
+        Bpmn.readModelFromStream(
+            new ByteArrayInputStream(
+                processDefinitionEntity.bpmnXml().getBytes(StandardCharsets.UTF_8)));
```

Comment:
For this first iteration we agreed to not add any caching at this point. As this will be updated to also query process instance data in #28432 we can address that later.

Reply:
Thanks, I found that one while looking at the flow node instance API, but could not use it in the current state mainly because it currently only caches flow node names. I would have planned to first finish #28432 and then look into how this could be properly cached.

Reply:
That makes sense, I just wanted to make sure you're aware 👍